libs/pando/engine/src/optimization/prune.ts

```
@@ -0,0 +1,566 @@
+export function prune<I extends OP, C extends Component>(
+  nodes: NumNode<I>[],
+  builds: C[][],
+  cat: string,
+  minimum: number[],
+  _keepTop: number
+): { nodes: NumNode<I>[]; builds: Omit<C, string>[][]; minimum: number[] } {
+  const state = new State(nodes, builds, cat)
+  while (state.progress) {
+    state.progress = false
+    pruneBranches(state)
+    minimum = pruneRange(state, minimum)
```

Biome (1.9.4)

[error] Reassigning a function parameter is confusing.

The parameter is declared here:

Use a local variable instead.

(lint/style/noParameterAssign)